openTSNE/affinity.py

```diff
@@ -129,14 +133,27 @@ def __init__(
         n_jobs=1,
         random_state=None,
         verbose=False,
+        k_neighbors="auto",
     ):
         self.n_samples = data.shape[0]
+        
+        if k_neighbors=="auto":
+            self._k_neighbors = min(self.n_samples - 1, int(3 * perplexity))
+        else:
+            self._k_neighbors = k_neighbors
```

Comment:
If we agree that the `k` parameter in `transform` should be completely independent of the number of nearest neighbors used to construct the embedding, I don't see a reason to store this on `self`. It just pollutes the class with unnecessary attributes. A local variable should be fine here.

Reply:
I am using this variable in `check_perplexity()`...

Reply:
Oh, that's right! In this case, we could potentially allow increasing the perplexity! For instance, if we construct the Affinity model with perplexity=30 and k=300, calling `set_perplexity(100)` should work fine. So there still shouldn't be any dependence on `self._k_neighbors`. The dependence should be on `self.__neighbors.shape[1]`.

The key thing with `set_perplexity` was to not recompute the distance matrix. Everything that we can do with the existing distance matrix should be allowed.

Come to think of it, calling `set_perplexity(300)` should technically also work fine, since we're allowing that in the constructor. This is getting confusing. However, if we instantiate `a = Aff(perplexity=100, k_neighbors=300)`, and run that for a bit, then call `a.set_perplexity(250)` and run that for a bit. The embedding won't change much, since we're still working with the same KNNG. Which will be even more confusing.

However, I think consistency is more important, so if we allow `perplexity=250, k=300` in the constructor, this should definitely also be allowed using `set_perplexity`.

Reply:
I think my current implementation allows that. But adding a warning is a good idea. 

To be honest, I am not a fan of `check_perplexity()` setting perplexity automatically if the provided value is too high. Why not simply exit with error?

Reply:
Oh, I missed the change in `check_perplexity`. That's not okay then, because `to_new` also calls `check_perplexity`, and it would be dependent on the original `k_neighbors`, which we don't want.

The purpose of `check_perplexity` is to make sure that the perplexity is within a valid range. With this change, we're allowing perplexity to be anything from 1 to K. And K should be anything between 1 to N. So maybe it would make more sense to change the function to accept both k and perplexity as parameters, and check if they're within these bounds. Then, in standard t-SNE, we'd call them using `check_perplexity(perplexity, N)`, and in transform `check_perplexity(transform_perplexity, N_reference_samples)`.

Reply:
OK sounds good. I tried to implement this.